fix(tests): test_reingest_from_s3 prefix tests fail with KeyError after #4421 prepass added

## Problem

5 tests in `packages/scraper-framework/tests/test_reingest_from_s3.py` fail on every PR that triggers the `ingestion-tests` CI job. The failures all share the shape `KeyError: <MagicMock id='...'>` raised at `scripts/reingest_from_s3.py:3952` inside the new judge pre-pass added by PR #4421 (`feat(reingest): apply judge pre-pass to run_reingest_from_prefix`).

```
FAILED tests/test_reingest_from_s3.py::TestRunReingestFromPrefixBustCachePropagation::test_propagates_true_to_each_submit - KeyError: <MagicMock id='139689888703424'>
FAILED tests/test_reingest_from_s3.py::TestRunReingestFromPrefixBustCachePropagation::test_default_is_false - KeyError: <MagicMock name='mock.submit()' id='139689889598352'>
FAILED tests/test_reingest_from_s3.py::TestRunReingestFromPrefix::test_processes_documents_with_pool - KeyError: <MagicMock id='139879170954144'>
FAILED tests/test_reingest_from_s3.py::TestRunReingestFromPrefix::test_summary_reports_hash_mismatch_warnings - KeyError: <MagicMock id='139879094040288'>
FAILED tests/test_reingest_from_s3.py::TestRunReingestFromPrefix::test_summary_no_warning_when_all_hashes_match - KeyError: <MagicMock id='139879094521344'>
```

## Why this slipped through

PR #4421 only modified files under `scripts/`, so the path-filter-conditional `ingestion-tests` job in `.github/workflows/ci.yml` was *skipped* on its CI run. Main CI on subsequent commits also skips `ingestion-tests` whenever the merged commit doesn't touch `packages/scraper-framework/`. The regression is invisible on main but reliably trips any PR that does touch the package — including #4445 (this issue's parent #4437).

## Root cause

PR #4421 added a new judge pre-pass (lines 3940-3963 in scripts/reingest_from_s3.py) that uses `ThreadPoolExecutor` and `as_completed()` to fan out S3 fetches before the main `ProcessPoolExecutor` block:

```python
with ThreadPoolExecutor(max_workers=concurrency) as pool:
    futures = {
        pool.submit(_fetch_s3_content, s3_client, bucket, target["s3_key"]): idx
        for idx, target in enumerate(scan_targets)
    }
    for future in as_completed(futures):
        idx = futures[future]   # ← line 3952, KeyError raised here
```

The pre-existing tests in `test_reingest_from_s3.py` patch `reingest_from_s3.as_completed` with a `return_value=[future1, future2]` — those futures are scoped to the *outer* `ProcessPoolExecutor` block. When the pre-pass runs first, its inner `as_completed(futures)` call returns the test's patched list (because the patch replaces the global `as_completed` symbol used by both code paths), but those futures are NOT keys in the pre-pass's local `futures` dict, so `idx = futures[future]` raises `KeyError`.

## Fix options

1. **Patch the prepass to use a local `as_completed` lookup** that handles unknown futures gracefully (skip them) — defends the prepass against the test mock pattern but is also defense-in-depth in production.
2. **Update the affected tests** to either (a) also patch the inner ThreadPoolExecutor and `as_completed` separately, or (b) early-return from the prepass when scan_targets is empty (set up the test fixtures so prepass is a no-op).
3. **Have the prepass take an injected executor + as_completed** for testability.

Option 2(b) is probably the cheapest — pass `--skip-judge-prepass=True` from the test (the prepass skip flag exists per the docstring on `_expand_single_word_judge_surname`), or arrange the fixtures so `scan_targets` is empty. Looking at the prepass code, it skips when `not scan_targets`, so configuring `_list_s3_keys` / `_discover_courts` to return an empty/non-matching set should bypass the pool entirely.

## Acceptance Criteria

- [ ] All 5 failing tests in `test_reingest_from_s3.py::TestRunReingestFromPrefix*` pass after the fix.
  - Verify: `packages/scraper-framework/.venv/bin/pytest packages/scraper-framework/tests/test_reingest_from_s3.py::TestRunReingestFromPrefix packages/scraper-framework/tests/test_reingest_from_s3.py::TestRunReingestFromPrefixBustCachePropagation -v` exits 0.
- [ ] Existing pre-pass behavior is unchanged on real (non-mocked) inputs — the fix should not weaken the prepass's production behavior.
  - Verify: spot-check the prepass code path against a real S3 client in dev (e.g. via the existing reingest invocation).
- [ ] Add a regression test that fails on the original PR #4421 SHA (or a comment in the test file documenting the slip-through pattern) so the same shape can't recur silently.
  - Verify: the test or comment cites `ingestion-tests` skip and the prepass-vs-mainpool boundary explicitly.

## Why p1

Every PR that touches `packages/scraper-framework/` is currently blocked by these tests. That includes scraper fixes, ingestion bug fixes, and resolver tweaks — all the "data correctness" follow-up work. Unblocking is workflow-critical.

## Context

Discovered while landing PR #4445 for issue #4437 (San Diego NULL-judge fix). Diff there is a 30-line code change in `packages/scraper-framework/src/ingestion/db.py` plus 4 regression tests; locally `pytest tests/` passes for the package (7608 passed). CI's `ingestion-tests` job re-runs the same test suite under the path-filtered conditional and trips on the unrelated 5 failures above.

--- a/packages/scraper-framework/tests/test_reingest_from_s3.py
+++ b/packages/scraper-framework/tests/test_reingest_from_s3.py
@@ -12146,12 +12146,25 @@ class TestRunReingestFromPrefix:
         future2.result.return_value = "ok"
         pool.submit.side_effect = [future1, future2]
 
-        # Mock as_completed to return both futures
+        # Mock as_completed to return both futures.
+        #
+        # ``skip_judge_prepass=True`` bypasses the #4419 judge pre-pass.
+        # Without it the test trips ``KeyError: <MagicMock>`` at the prepass's
+        # ``idx = futures[future]`` line: the global ``as_completed`` patch
+        # below returns the outer-pool futures, but the prepass passes its
+        # own local ``futures`` dict (whose keys are
+        # ``ThreadPoolExecutor.submit(_fetch_s3_content, ...)`` futures, not
+        # the outer-pool ones) to that same patched ``as_completed``.  See
+        # #4449 for the slip-through-via-``ingestion-tests``-skip diagnosis.
+        # This test exercises the main ``ProcessPoolExecutor`` flow only —
+        # the prepass has dedicated coverage in
+        # ``TestRunReingestFromPrefixJudgePrepassBoundary``.
         with patch("reingest_from_s3.as_completed", return_value=[future1, future2]):
             stats = reingest.run_reingest_from_prefix(
                 "postgresql://test",
                 prefix="federal/",
                 concurrency=4,
+                skip_judge_prepass=True,
             )
 
         assert stats["total_keys"] == 2
@@ -12260,6 +12273,8 @@ class TestRunReingestFromPrefix:
         pool.submit.side_effect = [future1, future2, future3]
 
         mock_logger = MagicMock()
+        # ``skip_judge_prepass=True`` — see comment in
+        # ``test_processes_documents_with_pool`` and #4449.
         with (
             patch(
                 "reingest_from_s3.as_completed",
@@ -12271,6 +12286,7 @@ class TestRunReingestFromPrefix:
                 "postgresql://test",
                 prefix="federal/",
                 concurrency=4,
+                skip_judge_prepass=True,
             )
 
         assert stats["total_keys"] == 3
@@ -12343,6 +12359,8 @@ class TestRunReingestFromPrefix:
         pool.submit.side_effect = [future1, future2]
 
         mock_logger = MagicMock()
+        # ``skip_judge_prepass=True`` — see comment in
+        # ``test_processes_documents_with_pool`` and #4449.
         with (
             patch(
                 "reingest_from_s3.as_completed",
@@ -12354,6 +12372,7 @@ class TestRunReingestFromPrefix:
                 "postgresql://test",
                 prefix="federal/",
                 concurrency=4,
+                skip_judge_prepass=True,
             )
 
         assert stats["hash_mismatch_warnings"] == 0
@@ -13915,12 +13934,15 @@ class TestRunReingestFromPrefixBustCachePropagation:
         future2.result.return_value = {"status": "ok", "hash_mismatch": False}
         pool.submit.side_effect = [future1, future2]
 
+        # ``skip_judge_prepass=True`` — see comment in
+        # ``test_processes_documents_with_pool`` and #4449.
         with patch("reingest_from_s3.as_completed", return_value=[future1, future2]):
             reingest.run_reingest_from_prefix(
                 "postgresql://test",
                 prefix="federal/",
                 concurrency=4,
                 bust_llm_cache=True,
+                skip_judge_prepass=True,
             )
 
         assert pool.submit.call_count == 2
@@ -13981,16 +14003,216 @@ class TestRunReingestFromPrefixBustCachePropagation:
         future.result.return_value = {"status": "ok", "hash_mismatch": False}
         pool.submit.return_value = future
 
+        # ``skip_judge_prepass=True`` — see comment in
+        # ``test_processes_documents_with_pool`` and #4449.
         with patch("reingest_from_s3.as_completed", return_value=[future]):
             reingest.run_reingest_from_prefix(
                 "postgresql://test",
                 prefix="federal/",
                 concurrency=1,
+                skip_judge_prepass=True,
             )
 
         assert pool.submit.call_count == 1
         call = pool.submit.call_args_list[0]
         assert call.args[6] is False
+
+
+class TestRunReingestFromPrefixJudgePrepassBoundary:
+    """Regression coverage for #4449 — the #4419 judge pre-pass added to
+    ``run_reingest_from_prefix`` must not collide with tests that mock the
+    main ``ProcessPoolExecutor``'s ``as_completed`` globally.
+
+    Slip-through pattern (documented so the same shape can't recur silently):
+
+    * PR #4421 added a new ``ThreadPoolExecutor`` + ``as_completed`` block
+      to seed judges before the main ``ProcessPoolExecutor`` runs (lines
+      ~3940-3963 in ``scripts/reingest_from_s3.py``).
+    * The pre-pass's local ``futures`` dict has different keys than the
+      main pool's local ``futures`` dict.
+    * Five tests in ``TestRunReingestFromPrefix*`` patch
+      ``reingest_from_s3.as_completed`` globally with ``return_value=[<main
+      pool's futures>]``.  When the prepass calls ``as_completed`` on its
+      own local ``futures`` dict, it gets the test's main-pool futures back
+      and ``idx = futures[future]`` raises ``KeyError`` (line 3952).
+    * The regression hid in main CI because the path-filter-conditional
+      ``ingestion-tests`` job in ``.github/workflows/ci.yml`` is skipped on
+      every PR/main commit that doesn't touch ``packages/scraper-framework/``.
+      PR #4421 only modified ``scripts/reingest_from_s3.py``, so its CI run
+      skipped these tests entirely.
+
+    Tests in this class exercise the prepass-vs-main-pool boundary directly
+    by mocking ``_seed_judges_from_keys`` itself rather than the global
+    ``as_completed``.  They would have caught #4449 at PR-time on PR #4421.
+    """
+
+    @patch.dict(
+        os.environ,
+        {
+            "JUDGEMIND_ARCHIVE_BUCKET": "test-bucket",
+            "REDIS_URL": "redis://localhost:6379",
+            "OPENSEARCH_URL": "",
+        },
+    )
+    @patch("reingest_from_s3.boto3")
+    @patch("reingest_from_s3._list_s3_keys")
+    @patch("reingest_from_s3._seed_courts")
+    @patch("reingest_from_s3._discover_courts")
+    @patch("reingest_from_s3._seed_judges_from_keys")
+    @patch("reingest_from_s3.psycopg")
+    @patch("reingest_from_s3.ProcessPoolExecutor")
+    def test_prepass_runs_by_default_without_keyerror(
+        self,
+        mock_pool_cls: MagicMock,
+        mock_psycopg: MagicMock,
+        mock_seed_judges: MagicMock,
+        mock_discover: MagicMock,
+        mock_seed: MagicMock,
+        mock_list: MagicMock,
+        mock_boto3: MagicMock,
+    ) -> None:
+        """Default behavior (``skip_judge_prepass=False``): the prepass
+        is invoked and the run completes without ``KeyError``.
+
+        This test mocks ``_seed_judges_from_keys`` at the function boundary
+        (rather than mocking ``as_completed`` globally) so it does NOT
+        interact with the prepass's internal ``ThreadPoolExecutor`` state.
+        On the PR #4421 SHA without the test fix in this PR, every
+        ``run_reingest_from_prefix`` call that mocks ``as_completed``
+        globally (without bypassing this boundary) trips
+        ``KeyError: <MagicMock>`` at line 3952.
+        """
+        keys = ["federal/federal/courtlistener/raw/aaa111.html"]
+        mock_list.return_value = keys
+        mock_discover.return_value = [
+            {
+                "state": "Federal",
+                "county": "Federal",
+                "court_name": "Courtlistener, County of Federal",
+                "court_code": "federal-federal",
+                "timezone": "America/Los_Angeles",
+            }
+        ]
+        conn_mock = MagicMock()
+        mock_psycopg.connect.return_value.__enter__ = MagicMock(return_value=conn_mock)
+        mock_psycopg.connect.return_value.__exit__ = MagicMock(return_value=False)
+        mock_seed.return_value = {"federal-federal": "court-id-1"}
+
+        # Prepass returns its declared shape — no KeyError, no real S3
+        # fetch, no real DB write.
+        mock_seed_judges.return_value = {
+            "docs_scanned": 1,
+            "judges_seeded": 0,
+            "judges_skipped_invalid": 0,
+        }
+
+        pool = MagicMock()
+        mock_pool_cls.return_value.__enter__ = MagicMock(return_value=pool)
+        mock_pool_cls.return_value.__exit__ = MagicMock(return_value=False)
+
+        future = MagicMock()
+        future.result.return_value = {"status": "ok", "hash_mismatch": False}
+        pool.submit.return_value = future
+
+        with patch("reingest_from_s3.as_completed", return_value=[future]):
+            stats = reingest.run_reingest_from_prefix(
+                "postgresql://test",
+                prefix="federal/",
+                concurrency=1,
+                # skip_judge_prepass intentionally omitted — defaults to
+                # False so the prepass code path runs.
+            )
+
+        # Prepass was invoked exactly once, with the listed keys.
+        mock_seed_judges.assert_called_once()
+        call_kwargs = mock_seed_judges.call_args.kwargs
+        call_args = mock_seed_judges.call_args.args
+        # Positional arg layout: (conn, s3_client, keys, bucket, court_ids)
+        assert call_args[2] == keys, "prepass should receive the same keys list"
+        assert call_kwargs.get("concurrency") == 1
+        # Prepass stats are surfaced into the run summary.
+        assert stats["judge_prepass_docs_scanned"] == 1
+        assert stats["judge_prepass_judges_seeded"] == 0
+        assert stats["judge_prepass_judges_skipped_invalid"] == 0
+        # Main pool ran on top of the prepass without raising.
+        assert stats["total_keys"] == 1
+        assert stats["processed"] == 1
+
+    @patch.dict(
+        os.environ,
+        {
+            "JUDGEMIND_ARCHIVE_BUCKET": "test-bucket",
+            "REDIS_URL": "redis://localhost:6379",
+            "OPENSEARCH_URL": "",
+        },
+    )
+    @patch("reingest_from_s3.boto3")
+    @patch("reingest_from_s3._list_s3_keys")
+    @patch("reingest_from_s3._seed_courts")
+    @patch("reingest_from_s3._discover_courts")
+    @patch("reingest_from_s3._seed_judges_from_keys")
+    @patch("reingest_from_s3.psycopg")
+    @patch("reingest_from_s3.ProcessPoolExecutor")
+    def test_skip_judge_prepass_flag_bypasses_prepass(
+        self,
+        mock_pool_cls: MagicMock,
+        mock_psycopg: MagicMock,
+        mock_seed_judges: MagicMock,
+        mock_discover: MagicMock,
+        mock_seed: MagicMock,
+        mock_list: MagicMock,
+        mock_boto3: MagicMock,
+    ) -> None:
+        """``skip_judge_prepass=True`` must bypass ``_seed_judges_from_keys``
+        entirely.
+
+        This is the escape hatch the five sibling tests in
+        ``TestRunReingestFromPrefix`` and
+        ``TestRunReingestFromPrefixBustCachePropagation`` rely on to mock
+        ``as_completed`` globally without tripping the prepass's local
+        ``futures`` dict.  Verify the flag is honored so future refactors
+        don't silently re-introduce the prepass on the bypass path.
+        """
+        mock_list.return_value = [
+            "federal/federal/courtlistener/raw/aaa111.html",
+        ]
+        mock_discover.return_value = [
+            {
+                "state": "Federal",
+                "county": "Federal",
+                "court_name": "Courtlistener, County of Federal",
+                "court_code": "federal-federal",
+                "timezone": "America/Los_Angeles",
+            }
+        ]
+        conn_mock = MagicMock()
+        mock_psycopg.connect.return_value.__enter__ = MagicMock(return_value=conn_mock)
+        mock_psycopg.connect.return_value.__exit__ = MagicMock(return_value=False)
+        mock_seed.return_value = {"federal-federal": "court-id-1"}
+
+        pool = MagicMock()
+        mock_pool_cls.return_value.__enter__ = MagicMock(return_value=pool)
+        mock_pool_cls.return_value.__exit__ = MagicMock(return_value=False)
+
+        future = MagicMock()
+        future.result.return_value = {"status": "ok", "hash_mismatch": False}
+        pool.submit.return_value = future
+
+        with patch("reingest_from_s3.as_completed", return_value=[future]):
+            stats = reingest.run_reingest_from_prefix(
+                "postgresql://test",
+                prefix="federal/",
+                concurrency=1,
+                skip_judge_prepass=True,
+            )
+
+        # Prepass MUST NOT run when skip flag is set.
+        mock_seed_judges.assert_not_called()
+        # Prepass keys are absent from the summary when the prepass is skipped.
+        assert "judge_prepass_docs_scanned" not in stats
+        # Main pool still ran.
+        assert stats["total_keys"] == 1
+        assert stats["processed"] == 1
 
 
 class TestSplitChildGuardUntouched:

